File: app/src/main/java/ca/judacribz/gainzassist/async/FirebaseService.java
package ca.judacribz.gainzassist.async;

import android.app.IntentService;
import android.app.Service;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.widget.Toast;
import ca.judacribz.gainzassist.models.*;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;

import java.util.ArrayList;

import static ca.judacribz.gainzassist.firebase.Database.getWorkoutsRef;


public class FirebaseService extends IntentService implements ChildEventListener {

    private final static String SETS = "sets";

    WorkoutHelper workoutHelper = new WorkoutHelper(this);
    ArrayList<Exercise> exercises;
    ArrayList<Set> sets;

    Exercise exercise;
    Set set;

    WorkoutRepo workoutRepo;

    public FirebaseService() {
        super("FirebaseService");
    }

    @Override
    public void onCreate() {
        super.onCreate();

        workoutRepo = new WorkoutRepo(getApplication());
    }

    @Override
    public int onStartCommand(@Nullable Intent intent, int flags, int startId) {
        getWorkoutsRef().addChildEventListener(this);
        return Service.START_STICKY;
    }

    @Override
    protected void onHandleIntent(Intent intent) {
    }

    ArrayList<String> workoutNames;
    @Override
    public void onChildAdded(DataSnapshot workoutShot, String s) {
        workoutRepo.insertWorkout(extractWorkout(workoutShot));
//        workoutHelper.addWorkout(extractWorkout(workoutShot));
    }

    @Override
    public void onChildChanged(DataSnapshot workoutShot, String s) {

        workoutHelper.updateWorkout(workoutShot.getKey(), extractWorkout(workoutShot));

    }

    @Override
    public void onChildRemoved(DataSnapshot workoutShot) {
        Toast.makeText(this, "Deleted " + workoutShot.getKey(), Toast.LENGTH_SHORT).show();
        workoutHelper.deleteWorkout(workoutShot.getKey());

    }

    @Override
    public void onChildMoved(DataSnapshot dataSnapshot, String s) {

    }

    @Override
    public void onCancelled(DatabaseError databaseError) {

    }

    public Workout extractWorkout(DataSnapshot workoutShot) {
        exercises = new ArrayList<>();
        for (DataSnapshot exerciseShot : workoutShot.getChildren()) {
            // Add set to sets list
            sets = new ArrayList<>();
            for (DataSnapshot setShot : exerciseShot.child(SETS).getChildren()) {
                set = setShot.getValue(Set.class);

                if (set != null) {
                    set.setSetNumber(Integer.valueOf(setShot.getKey()));

                    sets.add(set);
                }
            }

            // Adds sets to exercise object, and add exercise to exercises list
            exercise = exerciseShot.getValue(Exercise.class);
            if (exercise != null) {
                exercise.setName(exerciseShot.getKey());
                exercise.setSets(sets);

                exercises.add(exercise);
            }
        }

        return new Workout(workoutShot.getKey(), exercises);
    }
}

File: app/src/main/java/ca/gainzassist/background/FirebaseService.kt
package ca.gainzassist.background

import android.app.IntentService
import android.content.Intent
import android.widget.Toast
import ca.gainzassist.domain.repository.WorkoutRepository
import ca.gainzassist.util.Misc.extractSession
import ca.gainzassist.util.Misc.extractWorkout
import ca.gainzassist.util.firebase.Database.getWorkoutSessionsRef
import ca.gainzassist.util.firebase.Database.getWorkoutsRef
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.orhanobut.logger.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

@Suppress("DEPRECATION")
class FirebaseService : IntentService("FirebaseService"), KoinComponent {

    private val workoutRepository: WorkoutRepository by inject()
    private val serviceScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var userWorkoutsRef: DatabaseReference? = null
    private var userSessionRef: DatabaseReference? = null

    private var workoutListener: ChildEventListener? = null
    private var sessionListener: ChildEventListener? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        userWorkoutsRef = getWorkoutsRef()
        userSessionRef = getWorkoutSessionsRef()

        if (workoutListener == null) {
            workoutListener = createWorkoutListener().also { listener ->
                userWorkoutsRef?.addChildEventListener(listener)
            }
        }

        if (sessionListener == null) {
            sessionListener = createSessionListener().also { listener ->
                userSessionRef?.addChildEventListener(listener)
            }
        }

        return START_STICKY
    }

    private fun createWorkoutListener(): ChildEventListener = object : ChildEventListener {
        override fun onChildAdded(workoutShot: DataSnapshot, s: String?) {
            val workout = extractWorkout(workoutShot)
            serviceScope.launch {
                workoutRepository.insertWorkout(workout, syncToFirebase = false)
            }
        }

        override fun onChildChanged(workoutShot: DataSnapshot, s: String?) {}

        override fun onChildRemoved(workoutShot: DataSnapshot) {
            Toast.makeText(this@FirebaseService, "Deleted " + workoutShot.key, Toast.LENGTH_SHORT).show()
        }

        override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {}

        override fun onCancelled(databaseError: DatabaseError) {
            Logger.d("FIREBASE DB WORKOUT ERROR: " + databaseError.message)
        }
    }

    private fun createSessionListener(): ChildEventListener = object : ChildEventListener {
        override fun onChildAdded(sessionShot: DataSnapshot, s: String?) {
            val session = extractSession(sessionShot)
            if (session != null) {
                serviceScope.launch {
                    workoutRepository.insertCompletedSession(session, syncToFirebase = false)
                }
            }
        }

        override fun onChildChanged(sessionShot: DataSnapshot, s: String?) {}

        override fun onChildRemoved(sessionShot: DataSnapshot) {}

        override fun onChildMoved(sessionShot: DataSnapshot, s: String?) {}

        override fun onCancelled(databaseError: DatabaseError) {
            Logger.d("FIREBASE DB SESSION ERROR: " + databaseError.message)
        }
    }

    override fun onHandleIntent(intent: Intent?) {}

    override fun onDestroy() {
        workoutListener?.let { listener ->
            userWorkoutsRef?.removeEventListener(listener)
        }
        sessionListener?.let { listener ->
            userSessionRef?.removeEventListener(listener)
        }
        workoutListener = null
        sessionListener = null
        userWorkoutsRef = null
        userSessionRef = null

        serviceScope.cancel()
        super.onDestroy()
    }
}
